### ISSUE TITLE
[Discussion & PR?] Smaller messages with "idstrategy = default" and "use-manifests = true"

### ISSUE THREAD

**Issue by guidomedina** [romix/akka-kryo-serialization#83]
@romix @luben I added some logging to debug my message sizes and for `idstrategy = "default"` and `use-manifests = true` my messages are smaller than with no manifest which makes think something must have changed?

What I mean here is that something must have changed that before no-manifest was rendering smaller messages.

Why is it `use-manifest = false` by default? Unless most people use `explicit` or `incremental` `idsestrategy`?, if that is not the case I propose two small changes, `use-manifests = true` by default and to invert the `if condition` where the first condition is `if (!useManifests) ...` negating the most probable execution.

I have tested Scala performance and in general (this is probably also true for Java), the most probable condition should be the first tested which accomplishes the best performance (less false positives or fallbacks to else) and would also make that `if condition` clearer _- performance here isn't compelling enough to change but easier to read -_

WDYT?

**Comment by guidomedina** [comment]
Why AkkaKryoCompression tests failed when by default `use-manifests = true`? I'm thinking to manually set `use-manifests = false` for that test config, maybe it expecting a size and it is another?

**Comment by guidomedina** [comment]
Yeah, I figured it out, for some tests the size has to match so you have to manually set `use-manifests = false` for these specific compression tests.

**Comment by guidomedina** [comment]
Do you think the current defaults at `reference.conf` are accurate? I ended with:
- `idstrategy = "default"`
- `type = "nograph"`
- `use-manifests = true`

My messages are classic DTO, some of them a bit long (10+ properties) and they range between 300 bytes to 1K which tells me 4K buffer is just over-killing for most cases, what was the criteria to pick these defaults?

**Comment by guidomedina** [comment]
Ignore the buffer-size comment, I see now you re-use the buffers (very smart) which is solely the reason why you pool KryoSerializer instead of just Kryo instances.

**Comment by luben** [comment]
@guidomedina the manifests are deprecated since scala 2.10. I would not make them default now. The compression (lz4) should be able to cope with extended size that comes from putting the fully qualified class names in the serialized messages.

**Comment by guidomedina** [comment]
@luben but the thing is there is a size difference between the two, I'm using Scala 2.11.8 with Java 8, my project is in Java so I'm sure if there is a possible impact there.

**Comment by guidomedina** [comment]
@luben when I try to use incremental or explicit I get the following, I have all my DTO classes registered, which is why I can only use "default", I'm not really sure how people are doing with "incremental" and "explicit", I see in the code they are both the same?

Note: To register this class use: kryo.register(java.util.ArrayList.class)

**Comment by luben** [comment]
@guidomedina , yes, you have to register also all the types used by your classes

**Comment by luben** [comment]
@guidomedina , regarding your previous comment, yes there will be a size difference because the fully qualified classname can be bigger than the serialized manifest

**Comment by guidomedina** [comment]
@luben I'm a bit lost about use-manifest, some tests depend on this to be false to pass and the Kryo method being call depends on it too.

**Comment by guidomedina** [comment]
@luben I'm seeing smaller size with "use-manifests = true"

**Comment by luben** [comment]
@guidomedina yes, because in this case you don't need to put the full classname in the serialized message

**Comment by guidomedina** [comment]
@luben isn't the class name put with "use-manifests = true"?

Let me be more explicit:

**Config 1:**
- idstrategy = "default"
- use-manifests = true
- kryo-reference-map = false
- type = "nograph"

**Config 2:**
- idstrategy = "default"
- use-manifests = false
- kryo-reference-map = false
- type = "nograph"

Config 1 is rendering smaller messages than Config 2. shouldn't be the opposite?

**Comment by luben** [comment]
Looked at the code. Forget all about the manifests I have written above - I was wrong. So

- when you have "use-manifests = false" you serialize also the object type in the message. 
- when you have "use-manifests = true"  you don't serialize it so when deserializing it you should specify the Class[Type] that you are going to read.

The relevant code is:
https://github.com/romix/akka-kryo-serialization/blob/master/src/main/scala/com/romix/akka/serialization/kryo/KryoSerializer.scala#L478

**Comment by guidomedina** [comment]
@luben so akka does it for you, but then you also do something else (or something less) based on "use-manifests = true" which for me renders a smaller size object:

This is what is making the difference: https://github.com/romix/akka-kryo-serialization/blob/master/src/main/scala/com/romix/akka/serialization/kryo/KryoSerializer.scala#L487

**Comment by guidomedina** [comment]
@luben and based on my previous comments I was wondering what is the general consensus on default parameters or quick hack for akka-kryo users, what do they really use, as for me, I find "incremental" and "explicit" setups annoying enough not to try them which pushed me back to default which with the aid of manifest (and not writing the class but the header FQCN) makes my objects smaller, around 30 bytes, which is around 15%+ of my message sizes hence, making reasonably compelling for me use.

**Comment by guidomedina** [comment]
My proposal is to change it to:

  ```
def toBinary(obj: AnyRef): Array[Byte] = {
    val buffer = getBuffer
    try {
      if (useManifests)
        kryo.writeObject(buffer, obj)
      else
        kryo.writeClassAndObject(buffer, obj)
      buffer.toBytes
    } finally
      releaseBuffer(buffer)
  }
```

And to make "use-manifests=true" as default, for the following reasons:

- Smaller size.
- Code readability (not negating useManifest)

**Comment by luben** [comment]
I am using it with explicit registration but I may want soon to change to the default registration. I am running some benchmark and tests now with default registration and using or not manifests

**Comment by guidomedina** [comment]
@luben remember that for akka-remote (which isn't great), size is the most important factor, so, few nanos won't beat high I/O

**Comment by guidomedina** [comment]
@luben and I have a branch ready with the changes if you decide to go for it, tests passing and all (few changes were required to re-adapt the tests to useManifests)

**Comment by luben** [comment]
Can you push the branch to your github repo to take a look. Currently with:

  idstrategy = "default"
  use-manifests = true

the deserialization test for immutable.LongMap fails (it's not the size I believe)

In my view the best option would be to be able to use the default id strategy but be able also to manually assign ids, something like the "incremental" but with fallback to the default instead of assiging a random id.

**Comment by guidomedina** [comment]
done, at my master branch.

**Comment by guidomedina** [comment]
@luben the reason some tests may have failed you is because some of them are hardcoded to use directly writeClassAndObject which assumes use-manifests = false, hence you have to change the config passed to a couple of tests, or remove hardcoded calls but remember my Scala knowledge is limited, in fact this is my 3rd Scala contribution ever ;-)

**Comment by guidomedina** [comment]
@luben I just force pushed to remove an extra space that reported an unnecessary change.

**Comment by luben** [comment]
I don't know why the LongMap test fails when using manifests. If the your suggestion was true (that the serializer assumes "use-manifests=false") it should fail also the immutable.HashMap tests that uses the same serializer (ScalaImmutableMapSerializer).

I looked at your branch. I don't like that we will be testing a non-default configuration.

**Comment by guidomedina** [comment]
@luben I can change that give me a minute.

**Comment by luben** [comment]
BTW, why do you want to change the default configuration? It will cause a lot of pain for the users when updating to next version.

**Comment by luben** [comment]
On the other side I like you cleanup of toBinary and fromBinary.

**Comment by guidomedina** [comment]
I see your point, I have changed my mind now, I shouldn't change the defaults but in the other hand there are some bad hardcoded calls there, see ScalaMapSerializers line 62, that is always assuming `use-manifests = false`

That is why the tests are failing...I can fix these weird assumptions without touching the config?

**Comment by guidomedina** [comment]
@luben maybe we should just get rid of use-manifests and let akka write meta-data on top of message?

**Comment by guidomedina** [comment]
@luben and use writeObject(...) call only (not writeClassAndObject(...))

**Comment by luben** [comment]
@guidomedina the manifest are only for the top-level type, they can't replace the serialized objects types inside the containers.

**Comment by luben** [comment]
More specifically, if you what to replace writeClassAndObject you should also replace readClassAndObject(obj) with readObject(obj, clazz), so where do you get the clazz from?

**Comment by guidomedina** [comment]
@luben I pushed again, only with minor corrections then, I think it is fair to say that the manifest usage needs a closer look, maybe there is something that can actually be improved there, I can only say that I'm getting smaller sizes with it = true.

**Comment by guidomedina** [comment]
@luben assuming use-manifests = true, we have the class already, there are two starting points to read and write and both pass the class, don't they?

**Comment by guidomedina** [comment]
@luben yeah I see they are special, I'm not too savvy with Kryo yet, I know there special serializers are registered for internal types and hence they should be deducted from the original main class?

Imagine I have a:

```
class A {
  public B b;
}

A a = new A()
```

I know what a is a "A" because of the manifest, I know that a.b is a "B" because it is part of A definition? That's all I know, I think some other expert has to jump in, these assumptions will reduce size indeed, as long as the link is satisfied, I cannot help any longer because of my lack of Scala.

**Comment by luben** [comment]
@guidomedina  We have the class for Map with erased type parameters so we don't know what are the classes of the keys and values and how to deserialize them.

On the previous comment, yes,  indeed... we have a room from improvement and clarification here.

In my view more interesting is to register the classes from the "mappings" and "classes" section even with the "default" strategy. This way it will use the optimized path for the registered classes but will fall-back to putting the FQCNs if it encounrers unregistered class. IMHO looks more useful than the "incremental" strategy.

@romix, do you have an opinion here? Should I change the behaviour of the "default" strategy, it looks to me compatible on the API level.

**Comment by guidomedina** [comment]
@luben cherry-pick last commit though, it doesn't change anything but fixes a potential uncleared buffer if an exception is thrown.

**Comment by luben** [comment]
Yes, looks good. Please send a PR, @romix can merge it.

**Comment by guidomedina** [comment]
@romix @luben I'm thinking pre-initialization code, to push N instances to the pool where N = available CPUs to avoid unpredictable initialization.

WDYT?

**Comment by guidomedina** [comment]
@romix @luben something like this inside `SerializerPool`:

  ```
  for (i <- 1 to Runtime.getRuntime.availableProcessors()) {
    add(newInstance())
  }
```

**Comment by luben** [comment]
@guidomedina lets not overcomplicate it. You should be able to do it from inside your appilcation.

**Comment by guidomedina** [comment]
How can you access the Kryo instance Akka created from Java? and should I call it after ActorSystem is up correct?

**Comment by luben** [comment]
yes, you are right. I am not sure how to make it.

**Comment by guidomedina** [comment]
@luben don't worry about, is not a big deal, I'll do some work and share it later, I'm thinking something like `Math.min(CPUs, remote-dispatchers.max_threads)`, but let's not over-complicate it for now, let me test the idea and I share it, maybe if you guys think it is too good not to have I send a PR.

**Comment by guidomedina** [comment]
Issue resolved by #85
